src/unwinder.rs

```diff
@@ -700,35 +774,147 @@ pub struct ModuleSvmaInfo {
+            b"__debug_frame" | b".debug_frame" => self.debug_frame.take(),
+            _ => None,
+        }
+    }
+    fn segment_file_range(&mut self, name: &[u8]) -> Option<Range<u64>> {
+        match name {
+            b"__TEXT" => self.text_segment_file_range.take(),
+            _ => None,
+        }
+    }
+    fn segment_data(&mut self, name: &[u8]) -> Option<D> {
+        match name {
+            b"__TEXT" => self.text_segment.take(),
+            _ => None,
+        }
+    }
+}
+
+#[cfg(feature = "object")]
+mod object {
+    use super::{ModuleSectionInfo, Range};
+    use object::read::{Object, ObjectSection, ObjectSegment};
+
+    impl<'data: 'file, 'file, O, D> ModuleSectionInfo<D> for &'file O
+    where
+        O: Object<'data, 'file>,
+        D: From<&'data [u8]>,
+    {
+        fn base_svma(&self) -> u64 {
+            if let Some(text_segment) = self.segments().find(|s| s.name() == Ok(Some("__TEXT"))) {
+                // This is a mach-O image. "Relative addresses" are relative to the
+                // vmaddr of the __TEXT segment.
+                return text_segment.address();
+            }
+
+            // For PE binaries, relative_address_base() returns the image base address.
+            // Otherwise it returns zero. This gives regular ELF images a base address of zero,
+            // which is what we want.
+            self.relative_address_base()
+        }
+
+        fn section_svma_range(&mut self, name: &[u8]) -> Option<Range<u64>> {
+            let section = self.section_by_name_bytes(name)?;
+            Some(section.address()..section.address() + section.size())
+        }
+
+        fn section_file_range(&mut self, name: &[u8]) -> Option<Range<u64>> {
+            let section = self.section_by_name_bytes(name)?;
+            let (start, size) = section.file_range()?;
+            Some(start..start + size)
+        }
+
+        fn section_data(&mut self, name: &[u8]) -> Option<D> {
+            let section = self.section_by_name_bytes(name)?;
+            section.data().ok().map(|data| data.into())
+        }
+
+        fn segment_file_range(&mut self, name: &[u8]) -> Option<Range<u64>> {
+            let segment = self.segments().find(|s| s.name_bytes() == Ok(Some(name)))?;
+            let (start, size) = segment.file_range();
+            Some(start..start + size)
+        }
+
+        fn segment_data(&mut self, name: &[u8]) -> Option<D> {
+            let segment = self.segments().find(|s| s.name_bytes() == Ok(Some(name)))?;
+            segment.data().ok().map(|data| data.into())
+        }
+    }
 }
 
 impl<D: Deref<Target = [u8]>> Module<D> {
     pub fn new(
         name: String,
         avma_range: std::ops::Range<u64>,
         base_avma: u64,
-        svma_info: ModuleSvmaInfo,
-        unwind_data: ModuleUnwindData<D>,
-        text_data: Option<TextByteData<D>>,
+        mut section_info: impl ModuleSectionInfo<D>,
```

Comment:
Hmm, this changes the API quite significantly, doesn't it? Now the consumers of this crate no longer decide themselves which type of unwind data gets used. In the past they could pick between `.debug_frame` and `.eh_frame`, for example. Now they just provide section contents and framehop will hopefully pick the right thing for them.

Reply:
In a way, yes, but if someone has a use case where they care about picking between those sections they can provide an appropriate impl to expose the things they are interested in. I think that use case is not very likely, but this interface still allows it nonetheless.